src/replica/HttpAsyncClient.cc

```diff
@@ -0,0 +1,399 @@
+using namespace std::placeholders;
+namespace asio = boost::asio;
+namespace http = boost::beast::http;
+
+namespace {
+
+LOG_LOGGER _log = LOG_GET("lsst.qserv.replica.HttpAsyncClient");
+
+http::verb method2verb(string const& method) {
+    if (method == "GET") return http::verb::get;
+    else if (method == "POST") return http::verb::post;
+    else if (method == "PUT") return http::verb::put;
+    else if (method == "DELETE") return http::verb::delete_;
+    throw invalid_argument("HttpAsyncClient::" + string(__func__) + " invalid method '" + method + "'.");
+}
+} /// namespace
+
+namespace lsst {
+namespace qserv {
+namespace replica {
+
+string HttpAsyncClient::state2str(State state) {
+    switch (state) {
+        case State::CREATED: return "CREATED";
+        case State::IN_PROGRESS: return "IN_PROGRESS";
+        case State::FINISHED: return "FINISHED";
+        case State::FAILED: return "FAILED";
+        case State::BODY_LIMIT_ERROR: return "BODY_LIMIT_ERROR";
+        case State::CANCELLED: return "CANCELLED";
+        case State::EXPIRED: return "EXPIRED";
+    }
+    throw invalid_argument(
+            "HttpAsyncClient::" + string(__func__) + " unknown state: "
+            + to_string(static_cast<int>(state)) + ".");
+}
+
+
+shared_ptr<HttpAsyncClient> HttpAsyncClient::create(
+        asio::io_service& io_service, CallbackType const& onFinish,
+        string const& method, string const& url, string const& data,
+        std::map<std::string, std::string> const& headers,
+        size_t maxResponseBodySize,
+        unsigned int expirationIvalSec) {
+    return shared_ptr<HttpAsyncClient>(new HttpAsyncClient(
+            io_service, onFinish, method, url, data, headers, maxResponseBodySize, expirationIvalSec));
+}
+
+
+HttpAsyncClient::HttpAsyncClient(
+        asio::io_service& io_service, CallbackType const& onFinish,
+        string const& method, string const& url, string const& data,
+        std::map<std::string, std::string> const& headers,
+        size_t maxResponseBodySize,
+        unsigned int expirationIvalSec)
+    :   _io_service(io_service),
+        _resolver(io_service),
+        _socket(io_service),
+        _onFinish(onFinish),
+        _method(::method2verb(method)),
+        _url(url),
+        _data(data),
+        _headers(headers),
+        _maxResponseBodySize(maxResponseBodySize),
+        _expirationIvalSec(expirationIvalSec),
+        _expirationTimer(io_service),
+        _timer(io_service) {
+}
+
+
+HttpAsyncClient::~HttpAsyncClient() {
+    _expirationTimer.cancel();
+    _timer.cancel();
+    boost::system::error_code ec;
+    _socket.shutdown(asio::ip::tcp::socket::shutdown_both, ec);
+}
+
+
+void HttpAsyncClient::start() {
+    string const context = "HttpAsyncClient::" + string(__func__) + " ";
+    util::Lock lock(_mtx, context);
+    _assertState(lock, context, {State::CREATED});
+    try {
+        _parsUrl(lock);
+        _prepareRequest(lock);
+        _resolve(lock);
```

Comment:
I agree regarding the first two methods. It does make sense to move both to the c-tor. The third method should be initiated here because once created, the request object might be stored in some queue and be initiated at the right time (once it's needed). That was the conscious decision to separate object construction from any network operations. Hence its has two states in its FSA:
```
CREATE -> IN_PROGRESS -> 
```